### ISSUE TITLE
`assert(self.attribute, isinstance(...))` information does not fully propagate across methods

### ISSUE THREAD

**Issue by dzhu** [python/mypy#7327]
This is a bug report.

* `python 3.7.4`
* `mypy 0.720` and `0.730+dev.6e75f9d208642dacb9b4d73cdd9dd319c88a4c70`

Consider this code (let's call it `test.py`):
```python
class Base:
    def __init__(self) -> None:
        self.base_attr = None


class Derived(Base):
    def __init__(self) -> None:
        super().__init__()
        self.derived_attr = None


class C:
    def __init__(self, x: Base) -> None:
        self.x = x
        assert isinstance(self.x, Derived)
        self.y = self.x

    def go(self) -> None:
        self.x.derived_attr  # line 19
        self.y.derived_attr
```

The full output of `mypy test.py` is as follows:
```
test.py:19: error: "Base" has no attribute "derived_attr"
```
That is, Mypy is able to deduce from the `assert isinstance` that `y` is of type `Derived` in `go` but cannot do the same for `x`, though it seems the same logic should apply.

There is no error if everything is put into one method (e.g., by commenting the `def go` line).

(In this exact case, we could simply annotate the argument with `x: Derived` or put the `assert` before the assignment, but this is simplified from something where the attribute is set in the superclass's `__init__` and the expected type varies across subclasses, so that wouldn't work there.)

**Comment by dzhu** [comment]
Well, plenty of information is definitely propagating across methods; my question was why `x` and `y` end up being treated differently in `go`. (Or, in the context of your example -- which makes things more obvious -- why `c.y = Base()` is an error and `c.x = Base()` is not.) It seems to not be about methods directly, but rather the distinction between type restrictions on values and those on attributes. The `assert` means that the _value_ of `self.x` is known to be of type `Derived`, but that only applies for the rest of the method: the inferred type for the _attribute_ `self.x` has already been set to `Base` and does not change. On the other hand, `self.y` is treated as `Derived` as both a value and an attribute, because the restriction is known during its initial assignment, so that survives outside the method. Is that anywhere near accurate? If so, then perhaps this is n the question becomes: is there any way to influence the deduced type restriction on an attribute from somewhere besides its initial assignment?

Having thought about the situation that way, I realized that it can be resolved by adding the annotation `x: Derived` in the class body. It would be helpful to have something that also works in Python 3.5, though.

**Comment by ilevkivskyi** [comment]
> Is that anywhere near accurate?

Yes, the type of variable/attribute is inferred from the first assignment (unless explicitly declared).
> It would be helpful to have something that also works in Python 3.5, though.

You can use `# type:` comments to annotate functions and/or variables on Python versions that don't support annotations (it is actually in one of the first examples in README, also reading the [docs](https://mypy.readthedocs.io/en/latest/type_inference_and_annotations.html#explicit-var-types) is often helpful).

**Comment by dzhu** [comment]
I am quite aware of that style of annotation, thank you; perhaps I should've mentioned that. However, there doesn't seem to be a way to use it to actually modify the already-deduced annotation in the example I gave, as I was mostly trying to do. I tried inserting all of these after `self.y = self.x`, and none help (in fact, the first two cause errors of their own):

```python
        self.x = self.x  # type: Derived
        self.x = self.y  # type: Derived
        self.x  # type: Derived
```

I did do something silly, though: when I mentioned that using `x: Derived` works, that was actually in a subclass of `C`, rather than `C` itself, to more closely match my full code. (Including it in `C` would just lead to other errors, of course.) I then completely forgot to try the last combination of comment-style annotations in a subclass. Turns out that does work! There's still some fiddling to do; a class-level annotation can't be used, but it works to have
```python
        self.x = self.x  # type: Derived
```
in the subclass `__init__`. So I guess there's another subtle behavior here where annotations can be used to overwrite deduced types from _superclasses_, but not the _same class_, or something like that.

Anyway, thanks for your help!
